internal/credentials/spiffe/spiffe.go

```diff
@@ -81,3 +83,29 @@ func BundleMapFromBytes(bundleMapBytes []byte) (BundleMap, error) {
 	}
 	return bundleMap, nil
 }
+
+// GetRootsFromSPIFFEBundleMap returns the root trust certificates from the
+// SPIFFE bundle map for the given trust domain from the leaf certificate.
+func GetRootsFromSPIFFEBundleMap(bundleMap BundleMap, leafCert *x509.Certificate) (*x509.CertPool, error) {
+	// 1. Upon receiving a peer certificate, verify that it is a well-formed SPIFFE
+	//    leaf certificate.  In particular, it must have a single URI SAN containing
+	//    a well-formed SPIFFE ID ([SPIFFE ID format]).
+	spiffeId := credinternal.SPIFFEIDFromCert(leafCert)
+	if spiffeId == nil {
+		return nil, fmt.Errorf("credinternal.SPIFFEIDFromCert(leafCert) = nil, failed to parse a SPIFFE id")
+	}
+
+	// 2. Use the trust domain in the peer certificate's SPIFFE ID to lookup
+	//    the SPIFFE trust bundle. If the trust domain is not contained in the
+	//    configured trust map, reject the certificate.
+	spiffeBundle, ok := bundleMap[spiffeId.Host]
+	if !ok {
+		return nil, fmt.Errorf("getRootsFromSPIFFEBundleMap() failed. No bundle found for peer certificates trust domain %v", spiffeId.Host)
```

Comment:
I think this error message will be bubbled up to the user, so can we make it a little more user friendly? For example, "peer leaf certificate contained a SPIFFE ID with an unrecognized trust domain (%s), but verification with a SPIFFE trust map was configured", and then insert the trust domain?

Reply:
Done